preps for Splunk APM layer (#13)
signalfx/lambda-layer-versions: splunk-apm/README.md

> # Splunk distribution of OpenTelemetry Lambda

Author: It will be :)

> Splunk OTEL lambda is a downstream distribution of opentelemetry-lambda. Supported OpenTelemetry Lambda layers are preconfigured for easy use with Splunk environment (ingest, SmartAgent). Users can enhance their existing Lambda functions by adding Splunk-managed layer directly. Layer ARNs [are published here](https://github.com/signalfx/lambda-layer-versions/).

Author: Partially accepted :)

## Get started 

Currently, following components are supported:
- [Java wrapper](#java-wrapper) 

> 

Author: I'd leave it here - "requirements" does not sound good to me in this context.

### Configuration

> All wrapper layers come preconfigured for quick start with Splunk.

Author: +1

1. Context propagation

>     By default, W3C specified Trace Context and Baggage (`tracecontext,baggage`) context propagation are used. 

Reviewer: ```suggestion
    By default, W3C specified Trace Context and Baggage (`tracecontext,baggage`) context propagators are used. 
```

Author: +1

>     By default, W3C specified Trace Context and Baggage (`tracecontext,baggage`) context propagation are used. 

Author: +1

>     If you want to change this, set the `OTEL_PROPAGATORS` environment variable in your Lambda function code. For more information about available context propagators, see the [Propagator settings](https://github.com/open-telemetry/opentelemetry-java/tree/v1.1.0/sdk-extensions/autoconfigure#customizing-the-opentelemetry-sdk) for the OpenTelemetry Java.

Reviewer 2: ```suggestion
    To change propagators, set the `OTEL_PROPAGATORS` environment variable in your Lambda function code. For more information about available context propagators, see the [Propagator settings](https://github.com/open-telemetry/opentelemetry-java/tree/v1.1.0/sdk-extensions/autoconfigure#customizing-the-opentelemetry-sdk) in the OpenTelemetry Java documentation.
```

2. Traces export
    By default, the OpenTelemetry Protocol (`otlp`) exporter is used.
    If you want to use this exporter, set these environment variables in your Lambda function code:
    ```
    OTEL_EXPORTER_OTLP_ENDPOINT=<Endpoint URL>
    SPLUNK_ACCESS_TOKEN="orgAccessToken"
    ```
    Splunk provides also token-authenticated `jaeger-thrift-splunk` exporter for customers that need to use that specific protocol. In order to use it, please set (example endpoint value for SmartAgent):
    ```
    OTEL_TRACES_EXPORTER=jaeger-thrift-splunk
    OTEL_EXPORTER_JAEGER_ENDPOINT=<Endpoint URL>
    SPLUNK_ACCESS_TOKEN="orgAccessToken"
    ``` 
    By default, `OTEL_EXPORTER_JAEGER_ENDPOINT` is set to `http://127.0.0.1:9080/v1/trace` (local SmartAgent endpoint)
   

>     Default span flush wait timeout is 1 second. It is controlled with a following property (value in milliseconds):

Reviewer: Everything else is environment variables and this is a property, is it something different from env vars? If it really is env var then replace property with env var and `:` with `=`

Author: It's also env var, thanks for spotting this.

> 2. Traces export
>     By default, the OpenTelemetry Protocol (`otlp`) exporter is used.
>     If you want to use this exporter, set these environment variables in your Lambda function code:
>     ```
>     OTEL_EXPORTER_OTLP_ENDPOINT=<Endpoint URL>
>     SPLUNK_ACCESS_TOKEN="orgAccessToken"
>     ```
>     Splunk provides also token-authenticated `jaeger-thrift-splunk` exporter for customers that need to use that specific protocol. In order to use it, please set (example endpoint value for SmartAgent):
>     ```
>     OTEL_TRACES_EXPORTER=jaeger-thrift-splunk
>     OTEL_EXPORTER_JAEGER_ENDPOINT=<Endpoint URL>
>     SPLUNK_ACCESS_TOKEN="orgAccessToken"
>     ``` 
>     By default, `OTEL_EXPORTER_JAEGER_ENDPOINT` is set to `http://127.0.0.1:9080/v1/trace` (local SmartAgent endpoint)
>    
>     Default span flush wait timeout is 1 second. It is controlled with a following property (value in milliseconds):
>     ```
>     OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT: 30000
>     ```

Reviewer 2: ```suggestion
2. Traces export
    This distribution uses the OpenTelemetry Protocol (`otlp`) exporter by default. 
    
    To use this exporter, set these environment variables in your Lambda function code:
    
    ```
    OTEL_EXPORTER_OTLP_ENDPOINT=<EndpointURL>
    SPLUNK_ACCESS_TOKEN="<orgAccessToken>"
    ```
    Splunk also provides a token-authenticated `jaeger-thrift-splunk` exporter for customers that require it. The following example shows how to set the environment variables pointing to the Smart Agent endpoint:

    ```
    OTEL_TRACES_EXPORTER=jaeger-thrift-splunk
    OTEL_EXPORTER_JAEGER_ENDPOINT=<Endpoint URL>
    SPLUNK_ACCESS_TOKEN="orgAccessToken"
    ``` 

  The default value of `OTEL_EXPORTER_JAEGER_ENDPOINT` is `http://127.0.0.1:9080/v1/trace`, which corresponds to the local Smart Agent endpoint.
   
    Default flush wait timeout for spans is 1 second. It can be controlled through the following property (value in milliseconds):

    ```
    OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT: 30000
    ```
```

Reviewer 2: Shouldn't we be moving away from Smart Agent instructions? Shouldn't we provide an Ingest endpoint example instead (the only valid use case for sending data using the Jaeger exporter, at least until support for OTLP http is gold).

3. OpenTelemetry Metrics export

    By default, OTEL metrics are disabled.    

>     

Reviewer 2: Why? Is it because they're not supported in this distro? Support is experimental for Java; shouldn't we explain how to enable it in case they want to try it out?

> 4. Sampling
> 
>     By default `always_on` sampler is used, meaning that all spans are always sampled.
> 

Reviewer 2: ```suggestion
4. Sampling

    `OTEL_TRACE_SAMPLER` is set to `always_on` by default, meaning that all spans are always sampled.

```

Author: +1

> 5. Logging
>     
>     Logging is controller with `OTEL_LAMBDA_LOG_LEVEL` env property. Default set to `WARN`. When set to `DEBUG`, the logging exporter is added to traces exporter in order to help verify exported data.

Reviewer 2: ```suggestion
5. Logging
    
    Logging is controlled by the `OTEL_LAMBDA_LOG_LEVEL` environment variable. Default is `WARN`. When set to `DEBUG`, the logging exporter is added to traces exporter in order to help verify exported data.
```

Reviewer 2: Could enabling `DEBUG` incur in increased costs in AWS? If that's the case, we should warn users.

Author: Good idea, will note this.

>     Logging is controller with `OTEL_LAMBDA_LOG_LEVEL` env property. Default set to `WARN`. When set to `DEBUG`, the logging exporter is added to traces exporter in order to help verify exported data.

Reviewer: ```suggestion
    Logging is controller with `OTEL_LAMBDA_LOG_LEVEL` environment variable. Default set to `WARN`. When set to `DEBUG`, the logging exporter is added to traces exporter in order to help verify exported data.
```

Author: +1

### Java wrapper

> Uses OpenTelemetry Java v1.5.0. 

Reviewer: Shouldn't it use splunk agent instead of otel? Are you going to update this version every time new agent is released? If you really need to use otel version here it might be better to replace it with 1.5.3 as 1.5.0 has a known bug https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases

Author: It does not use the agent - no autoinstrumentation in lambda (too slow unfortunately). Wrappers come directly from the SDK :)

> Uses OpenTelemetry Java v1.5.0. 
> 
> Main documentation of the upstream layer can be found [here.](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/java/README.md)
> 
> No code change is needed to instrument a lambda. However, the AWS_LAMBDA_EXEC_WRAPPER environment variable needs to be set accordingly: 
> - `/opt/otel-handler` - for wrapping regular handlers (implementing RequestHandler)
> - `/opt/otel-proxy-handler` - for wrapping regular handlers (implementing RequestHandler) proxied through API Gateway, enabling HTTP context propagation
> - `/opt/otel-stream-handler` - for wrapping streaming handlers (implementing RequestStreamHandler)

Reviewer 2: ```suggestion
The Java wrapper is based on OpenTelemetry Java version 1.5.0. 

The official documentation of the upstream layer can be found [here](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/java/README.md).

No code change is needed to instrument a lambda. However, the `AWS_LAMBDA_EXEC_WRAPPER` environment variable must be set accordingly: 

- `/opt/otel-handler` for wrapping regular handlers implementing `RequestHandler`
- `/opt/otel-proxy-handler` for wrapping regular handlers implementing `RequestHandler`, proxied through API Gateway, and enabling HTTP context propagation
- `/opt/otel-stream-handler` for wrapping streaming handlers implementing `RequestStreamHandler`
```

#### Configuration example

Assuming:
- Splunk APM direct ingest
- default context propagation (`tracecontext,baggage`)
- RequestHandler lambda
- US0 Splunk ingest

> Following env properties should be set:

Reviewer: ```suggestion
Following environment variables should be set:
```

Reviewer: Assuming env property is same as environment variable. Could also replace `:` with `=`

Reviewer: Actually if the conf file uses `:` then it might be better to use it. I'd pick one from `:` and `=` and use it everywhere for consistency.

Author: It's ":" - config is yaml :)

> Assuming:
> - Splunk APM direct ingest
> - default context propagation (`tracecontext,baggage`)
> - RequestHandler lambda
> - US0 Splunk ingest
> 
> Following env properties should be set:
> ```
> AWS_LAMBDA_EXEC_WRAPPER: /opt/otel-handler
> OTEL_TRACES_EXPORTER: jaeger-thrift-splunk
> SPLUNK_ACCESS_TOKEN: <token>
> OTEL_EXPORTER_JAEGER_ENDPOINT: https://ingest.us0.signalfx.com/v2/trace
> ```

Reviewer 2: ```suggestion

```
AWS_LAMBDA_EXEC_WRAPPER: /opt/otel-handler
OTEL_TRACES_EXPORTER: jaeger-thrift-splunk
SPLUNK_ACCESS_TOKEN: <token>
OTEL_EXPORTER_JAEGER_ENDPOINT: https://ingest.us0.signalfx.com/v2/trace
```

The example assumes the following:

- Data is sent to Splunk APM directly via Ingest endpoint.
- Context propagation is default (`tracecontext,baggage`).
- It's a RequestHandler lambda.
- Realm is `us0`.

Author: +1

## Troubleshooting

> 1. Traces not delivered 
>     - Set `OTEL_LAMBDA_LOG_LEVEL` to `DEBUG` and search for traces IDs in backend.
>     - Set `OTEL_LAMBDA_LOG_LEVEL` to `DEBUG` for `jaeger-thrift-splunk' exporter. This will show debug information for the exporter.
>     - Increase `OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT` value (in milliseconds) if backend / network is slow.

Reviewer 2: ```suggestion
If traces are not delivered:
    - Set `OTEL_LAMBDA_LOG_LEVEL` to `DEBUG` and search for traces IDs in the backend.
    - Set `OTEL_LAMBDA_LOG_LEVEL` to `DEBUG` for the `jaeger-thrift-splunk' exporter. This shows debug information for the exporter.
    - Increase `OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT`if the backend / network reacts slowly.
```

Author: +1
